hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/QueryResultProducerImpl.java

```diff
@@ -77,21 +75,36 @@ public ResultIterator<JetSqlRow> iterator() {
         return iterator;
     }
 
+    public void done() {
+        doneTracker.markDone();
+    }
+
     @Override
     public void onError(QueryException error) {
         assert error != null;
-        done.compareAndSet(null, error);
+        doneTracker.markDone(error);
     }
 
-    public void done() {
-        done.compareAndSet(null, NORMAL_COMPLETION);
+    public void consume(Inbox inbox) {
+        doneTracker.ensureNotDoneExceptionally();
+        produce(inbox);
     }
 
-    public void consume(Inbox inbox) {
-        ensureNotDone();
+    /**
+     * @return true if all results are produced to the Iterator.
+     */
+    public boolean produce(Inbox inbox) {
+        doneTracker.ensureNotDoneExceptionally();
+
+        // in case of close() being called
+        if (doneTracker.isDone()) {
+            inbox.clear();
+            return end();
+        }
         if (limit <= 0) {
-            done.compareAndSet(null, new ResultLimitReachedException());
-            ensureNotDone();
+            doneTracker.markDone();
+            inbox.clear();
```

Comment:
note to reviewers: I clean inbox, so that processor can go to complete() and end the job